clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java

```diff
@@ -35,14 +39,29 @@
     private final Login login;
     private final String serviceName;
     private final LoginType loginType;
+    private final SaslMechanism mechanism;
     private int refCount;
 
-    private LoginManager(LoginType loginType, Map<String, ?> configs) throws IOException, LoginException {
+    private LoginManager(LoginType loginType, SaslMechanism mechanism, Map<String, ?> configs) throws IOException, LoginException {
         this.loginType = loginType;
+        this.mechanism = mechanism;
         String loginContext = loginType.contextName();
-        login = new Login(loginContext, configs);
-        this.serviceName = getServiceName(loginContext, configs);
-        login.startThreadIfNeeded();
+        
+        switch (mechanism) {
+            case GSSAPI:
+                login = new KerberosLogin(loginContext, configs);
+                this.serviceName = getServiceName(loginContext, configs);
+                login.startThreadIfNeeded();
+                break;
+            case PLAIN:
+                login = new PlainLogin(loginContext, configs);
+                // serviceName is used as protocol in createSaslClient and needs to be non-null.
+                // For SASL/PLAIN, this does not need to be configurable
+                this.serviceName = "kafka";
+                break;
+            default:
+                throw new KafkaException("Unsupported SASL mechanism " + mechanism);
+        }
     }
 
     private static String getServiceName(String loginContext, Map<String, ?> configs) throws IOException {
```

Comment:
Have changed `getServiceName` to `findServiceName` as suggested by @ijuma above. Also rebased to the latest level.